apps/showcase/src/components/training/training.component.ts

```diff
@@ -0,0 +1,238 @@
+    NgComponentOutlet,
+    TrainingStepPresComponent,
+    NgbDropdown,
+    NgbDropdownToggle,
+    NgbDropdownMenu,
+    NgbDropdownItem
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  templateUrl: './training.component.html',
+  styleUrl: './training.component.scss'
+})
+export class TrainingComponent implements OnInit {
+  /**
+   * Index of the training step to display
+   */
+  public currentStepIndex = 0;
+  public showInstructions = true;
+  public showSolution = false;
+  /**
+   * Training steps that
+   */
+  public steps: WritableSignal<TrainingStep[]> = signal([]);
+  public stepNames = computed(() =>
+    this.steps().map((step) => step.description.title)
+  );
+  private readonly dynamicContentService = inject(DynamicContentService);
+
+  @Input()
+  public trainingPath = '';
+  @Input()
+  public title = '';
+
+  private async loadStepContent(step: TrainingStep) {
+    if (!step.dynamicContent.htmlContent()) {
+      void this.loadResource(step.description.htmlContentUrl).then((content) =>
+        step.dynamicContent.htmlContent.set(content)
+      );
+    }
+    const fileConfiguration = step.description.filesConfiguration;
+    if (fileConfiguration?.urls && !step.dynamicContent.project()?.files) {
+      await this.updateStepDynamicContent(step, Object.entries(fileConfiguration.urls).map(([path, url])=> ({path, url})));
+    }
+    if (fileConfiguration?.solutionUrls && !step.dynamicContent.solutionProject()?.files) {
+      await this.updateStepDynamicContent(step, [
+        ...Object.entries(fileConfiguration.urls).map(([path, url]) => ({path, url})),
+        ...Object.entries(fileConfiguration.solutionUrls).map(([path, url]) => ({path, url}))
+      ],
+      true);
+    }
+  }
+
+  private async loadSteps() {
+    const program: TrainingStepDescription[] = JSON.parse(await this.loadResource(`${this.trainingPath}/program.json`));
+    const stepsToLoad: TrainingStep[] = [];
+    program.forEach((desc) => {
+      const step = {
+        description: desc,
+        dynamicContent: {
+          htmlContent: signal(''),
+          project: signal(null),
+          solutionProject: signal(null)
+        }
+      };
+      void this.loadStepContent(step);
+      stepsToLoad.push(step);
+    });
+    this.steps.set(stepsToLoad);
+    const stepParam = window.location.href.match(new RegExp('#([0-9]+)'))?.[1];
+    const stepRequested = !stepParam || Number.isNaN(stepParam) ? 0 : Number(stepParam);
+    this.setCurrentStep(stepRequested);
+  }
+
+  private getFilesContent(resources: Resource[]) {
+    return resources.reduce((fileSystemTree: FileSystemTree, resource) => {
+      if (resource.path === './' || resource.path === '.') {
+        return {...fileSystemTree, ...JSON.parse(resource.content)} as FileSystemTree;
+      }
+      const sanitizedPath = resource.path.replace(new RegExp('^[.]/?'), '');
+      const parsedPath = sanitizedPath.split('/');
+      parsedPath.reduce((pointer, path, index) => {
```

Comment:
let's try to see if we ca replace that with jsonPath.apply?

Reply:
(future PR)